DeleteUserDataActivity/backupAndDelete.ts

```diff
@@ -0,0 +1,472 @@
+};
+
+/**
+ * Find all versions of a notification status, then backup and delete each document
+ * @param param0.notificationStatusModel instance of NotificationStatusModel
+ * @param param0.userDataBackup information about the blob storage account to place backup into
+ * @param param0.notification parent notification
+ *
+ */
+const backupAndDeleteNotificationStatus = ({
+  notificationStatusModel,
+  userDataBackup,
+  notification
+}: {
+  notificationStatusModel: NotificationStatusDeletableModel;
+  userDataBackup: IBlobServiceInfo;
+  notification: RetrievedNotification;
+}): TaskEither<DataFailure, readonly RetrievedNotificationStatus[]> => {
+  return executeRecursiveBackupAndDelete<RetrievedNotificationStatus>(
+    item =>
+      notificationStatusModel.deleteNotificationStatusVersion(
+        item.notificationId,
+        item.id
+      ),
+    userDataBackup,
+    item => `notification-status/${item.id}.json`,
+    notificationStatusModel.findAllVersionsByNotificationId(notification.id)
+  );
+};
+
+/**
+ * Backup and delete a given message
+ *
+ * @param param0.messageStatusModel instance of MessageStatusModel
+ * @param param0.userDataBackup information about the blob storage account to place backup into
+ * @param param0.message the message
+ */
+const backupAndDeleteMessage = ({
+  messageModel,
+  userDataBackup,
+  message
+}: {
+  messageModel: MessageDeletableModel;
+  userDataBackup: IBlobServiceInfo;
+  message: RetrievedMessageWithoutContent;
+}): TaskEither<DataFailure, RetrievedMessageWithoutContent> => {
+  return sequenceT(taskEitherSeq)<
+    DataFailure,
+    // tslint:disable-next-line: readonly-array
+    [
+      TaskEither<DataFailure, RetrievedMessageWithoutContent>,
+      TaskEither<DataFailure, string>
+    ]
+  >(
+    saveDataToBlob<RetrievedMessageWithoutContent>(
+      userDataBackup,
+      `message/${message.id}.json`,
+      message
+    ),
+    fromQueryEither(() =>
+      messageModel.deleteMessage(message.fiscalCode, message.id)
+    ).mapLeft(toDocumentDeleteFailure)
+  ).map(_ => message);
+};
+
+const backupAndDeleteMessageContent = ({
+  messageContentBlobService,
+  messageModel,
+  userDataBackup,
+  message
+}: {
+  messageContentBlobService: BlobService;
+  messageModel: MessageDeletableModel;
+  userDataBackup: IBlobServiceInfo;
+  message: RetrievedMessageWithoutContent;
+}): TaskEither<DataFailure, Option<MessageContent>> => {
+  return (
+    tryCatch(
+      () =>
+        messageModel.getContentFromBlob(messageContentBlobService, message.id),
+      toError
+    )
+      // type lift
+      // from TaskEither of Either of Option of X
+      // to TaskEither of X
+      // this way we collaps every left/none case into the same path
+      .chain(fromEither)
+      .chain(e => fromEither(fromOption(undefined)(e)))
+      .foldTaskEither<DataFailure, Option<MessageContent>>(
+        _ => {
+          // unfortunately, a document not found is threated like a query error
```

Comment:
Does this mean that the user may get a subset of his data￼?

Reply:
Yes, in case of an error while fetching data. The rationale is: we don't want to impede the operation in case of a missing blob (which is a case of inconsistence of data, and should never happen).